File: scratch/js/bin/mule.py
#!/usr/bin/env python3

import json
import math
import time
from datetime import datetime

with open("../../tmp/mcc.json", "r") as f:
    data = json.load(f)

rows = []


def emit(key, value):
    rows.append({"key": key, "value": value})


def view_map(doc: dict):
    if mcc := doc.get("mcc_cdc"):
        source: str = (
            doc["contentType"]["source"] if "contentType" in doc else "tagging"
        )
        emit(
            [
                datetime.fromtimestamp(mcc["sequence"]).isoformat(),
                mcc["sequence"],
                mcc["update_type"],
                mcc["object_type"],
                source,
            ],
            1,
        )


TOTAL = 12339702
scale = math.ceil(TOTAL / len(data))

print(f"Processing {len(data)} rows {scale} times each")
start = time.time()
for i in range(scale):
    for row in data:
        view_map(row)
duration = (time.time() - start) * 1000
print(f"Emitted {len(rows)} rows in {duration:.0f}ms")

File: scratch/js/bin/mule.js
const data = require("../../../tmp/mcc.json");

const rows = [];
function emit(key, value) {
  rows.push({ key, value });
}

const viewMap = function (doc) {
  if (!doc.mcc_cdc) {
    return;
  }
  const source = doc.contentType ? doc.contentType.source : "tagging";
  emit(
    [
      new Date(doc.mcc_cdc.sequence * 1000).toISOString(),
      doc.mcc_cdc.sequence,
      doc.mcc_cdc.update_type,
      doc.mcc_cdc.object_type,
      source,
    ],
    1,
  );
};

const TOTAL = 12339702;
const scale = Math.ceil(TOTAL / data.length);

console.log(`Processing ${data.length} rows ${scale} times each`);
const start = Date.now();
for (let i = 0; i < scale; i++) {
  for (const row of data) {
    viewMap(row);
  }
}
const duration = Date.now() - start;
console.log(`Emitted ${rows.length} rows in ${duration}ms`);
